src/Juvix/Core/IR/Typechecker.hs

```diff
@@ -134,57 +176,65 @@ cType p ii g (Lam s) ann =
         (cSubst 0 (Free (Local ii)) s) -- x (varType) in context S with sigma*pi usage.
         (sig, ty' (vfree (Local ii))) -- is of type M (usage sigma) in context T
     _ → throwError $ show (snd ann) <> " is not a function type but should be."
+--
 cType p ii g (Elim e) ann = do
   ann' ← iType p ii g e
   unless
     (fst ann == fst ann' && quote0 (snd ann) == quote0 (snd ann'))
     (throwError (errorMsg ii (Elim e) ann ann'))
 
 -- inferable terms have type as output.
-iType0 ∷ (Show primTy, Show primVal, Eq primTy, Eq primVal) ⇒ Parameterisation primTy primVal → Context primTy primVal → Elim primTy primVal → Result (Annotation primTy primVal)
+iType0 ∷
+  (Show primTy, Show primVal, Eq primTy, Eq primVal) ⇒
+  Parameterisation primTy primVal →
+  Context primTy primVal →
+  Elim primTy primVal →
+  Result (Annotation primTy primVal)
 iType0 p = iType p 0
 
 iTypeErrorMsg ∷ Natural → Name → String
 iTypeErrorMsg ii x =
-  "Cannot find the type of \n"
-    <> show x
-    <> "\n (binder number "
-    <> show ii
+  "Cannot find the type of \n" <> show x <> "\n (binder number " <> show ii
     <> ") in the environment."
 
-iType ∷ (Show primTy, Show primVal, Eq primTy, Eq primVal) ⇒ Parameterisation primTy primVal → Natural → Context primTy primVal → Elim primTy primVal → Result (Annotation primTy primVal)
+iType ∷
+  (Show primTy, Show primVal, Eq primTy, Eq primVal) ⇒
+  Parameterisation primTy primVal →
+  Natural →
+  Context primTy primVal →
+  Elim primTy primVal →
+  Result (Annotation primTy primVal)
 -- the type checker should never encounter a bound variable (as in LambdaPi)? To be confirmed.
 iType _ _ii _g (Bound _) = error "Bound variable cannot be inferred"
 iType _ ii g (Free x) =
   case lookup x g of
     Just ann → return ann
     Nothing → throwError (iTypeErrorMsg ii x)
--- Prim-Const typing rule
+-- Prim-Const and Prim-Fn, pi = omega
 iType p _ii _g (Prim prim) =
-  let arrow (x : []) = VPrimTy x
+  let arrow [x] = VPrimTy x
       arrow (x : xs) = VPi Omega (VPrimTy x) (const (arrow xs))
    in case typeOf p arrow prim of
         Right a → return (Omega, VPrimTy a)
         Left f → return (Omega, f)
--- App rule, function M applies to N
+-- App, function M applies to N (Elimination rule of dependent function types)
 iType p ii g (App m n) = do
   mTy ← iType p ii g m -- annotation of M is usage sig and Pi with pi usage.
   case mTy of
     (sig, VPi pi varTy resultTy) → do
-      cType p ii g n (pi, varTy) -- N has to be of type varTy with usage pi
+      cType p ii g n (sig <.> pi, varTy) -- N has to be of type varTy with usage sig*pi
```

Comment:
As per lang ref, N has to be of type varTy with usage sig*pi, not pi.

Reply:
Right, still TBD if we have the constraint of sigma = 1.